### ISSUE TITLE
Transformer modules should not hard-code the activation function

### ISSUE THREAD

**Issue by markrogersjr** [pytorch/pytorch#65622]
## 🚀 Feature
When passing a `str` to the `activation` argument, only `torch.nn.functional.relu` and `torch.nn.functional.gelu` activation functions are supported. The `activation` argument should be able to support any activation function defined in `torch.nn.functional`.

## Motivation

I want to be able to use the PyTorch Transformer `torch.nn.module`s with any activation function that is defined in PyTorch. Here is the list of `torch.nn.module`s:

- `torch.nn.Transformer`
- `torch.nn.TransformerEncoder`
- `torch.nn.TransformerEncoderLayer`
- `torch.nn.TransformerDecoder`
- `torch.nn.TransformerDecoderLayer`

The activation functions live in both `torch.nn` (e.g., `torch.nn.ReLU`, `torch.nn.SiLU`) and `torch.nn.functional` (e.g., `torch.nn.functional.relu`, `torch.nn.functional.silu`). Mathematically, any function which maps an input float `torch.Tensor` to an output `torch.Tensor` of the same shape that is defined in either of the above two namespaces (`torch.nn` and `torch.nn.functional`) should be supported.

## Pitch

Redefine `transformer.py:_get_activation_fn` as follows:

```python
def _get_activation_fn(activation):
    try:
        activation_fn = getattr(torch.nn.functional, activation)
    except AttributeError:
        try:
            activation_fn = getattr(torch.nn, activation)()
        except AttributeError:
            raise TypeError(f'Activation function {activation} is not defined in neither `torch.nn` nor `torch.nn.functional`')
    return activation_fn
```

## Alternatives

 - Create a dedicated module for activation functions and have `_get_activation_fn` interact with that
 - Hard-code the list of activation function names supported as opposed to using `getattr`
 - Instead of passing in a string `activation` to one of the above `torch.nn.module`s, pass in an activation function that has already been instantiated.

## Additional context

A major part of model development, performance tuning, experimentation, etc. is trying different activation functions. Often, a model is part of a larger system such as a hyperparameter tuner which compares activation functions. No hacking should be necessary to achieve this functionality.

Furthermore, as the field of machine learning matures, new activation functions will be discovered. This feature ensures forward-compatibility as long as PyTorch supports those new activation functions.

**Comment by jbschlosser** [comment]
Hey @markrogersjr, as of #61355 it is possible to pass any callable as the activation function for the Transformer modules. Examples:
```python
import torch
from torch import nn
import torch.nn.functional as F

# Pass a functional
m = torch.nn.Transformer(..., activation=F.silu)

# Pass a module
m = torch.nn.Transformer(..., activation=nn.LeakyReLU())

# Pass a lambda
m = torch.nn.Transformer(..., activation=lambda x: F.relu(x + 1))
```

**Comment by markrogersjr** [comment]
@jbschlosser yes, you can pass in an instantiated activation function, but `_get_activation_fn` is hardcoded to only accept `relu` and `gelu`

**Comment by soumith** [comment]
let's be a bit more precise here.

Callables are allowed as the activation function after #61355.

However, passing the string "relu" etc. as the activation is not generalized to the activation functions available in `torch.nn.functional`.

@markrogersjr opened a patch to fix this here: https://github.com/pytorch/pytorch/pull/65627

We have two paths forward.

1. As one of the maintainers of torch.nn, declare that callables are the only way forward, and that we don't want to expand the string-based arguments further (with a reasoning as to why)
2. Accept @markrogersjr 's patch, or a similar patch to allow this.

@jbschlosser you might have overlooked this smaller detail, please take a look and reply with what is the final decision.

**Comment by markrogersjr** [comment]
@soumith Thanks for the suggestions, I think both are reasonable!

**Comment by markrogersjr** [comment]
I believe Option 2 is the way to go. Here are some reasons why:
 - Option 1 would break code for those who use a `str` to specify `activation` once they upgrade PyTorch.
 - Option 2 makes code a little simpler than Option 1:
```python
# Option 1
activation = torch.nn.functional.relu
transformer = torch.nn.Transformer(activation=activation)

# Option 2
transformer = torch.nn.Transformer(activation='relu')
```
 - Option 2 is a bit more flexible and user-friendly. Sometimes, it's easier to pass in a `str`.

**Comment by albanD** [comment]
Hi,

Just to let you know that I added this issue as a topic for our next torch.nn meeting. It will unfortunately take a bit of time as some core people are on holidays right now. But we'll answer soon.

**Comment by markrogersjr** [comment]
@albanDI I appreciate you guys having an in-depth discussion on the issue. I'll keep an eye out for updates, thanks.

**Comment by albanD** [comment]
Hi,

Following discussions, we feel that option 1 is better here.
Note that we don't think that removing the existing string arguments is warranted though (as you mentioned that would be BC-breaking) but we do not plan on adding more support here.

The main problem with the string approach is that we will either have to maintain a set of activation string (which will go stale) or have overly flexible code that can do pretty surprising thing if the input string is not something that we expect (what happens in I pass "utils" in your PR?).

Also it is easy to architecture your code example above to make both option as simple. For example by importing it like `from torch.nn.functional import silu`.
